prov/mrail/src/mrail_init.c

```diff
@@ -165,30 +181,77 @@ int mrail_get_core_info(uint32_t version, const char *node, const char *service,
 		FI_DBG(&mrail_prov, FI_LOG_CORE,
 		       "--- Begin fi_getinfo for rail: %zd ---\n", i);
 
-		ret = fi_getinfo(version, NULL, NULL, OFI_GETINFO_INTERNAL, core_hints, &info);
+		ret = fi_getinfo(version, NULL, NULL, OFI_GETINFO_INTERNAL, core_hints, &rail_info[i]);
 
 		FI_DBG(&mrail_prov, FI_LOG_CORE,
 		       "--- End fi_getinfo for rail: %zd ---\n", i);
 		if (ret)
 			goto err;
 
-		if (!fi)
-			*core_info = info;
-		else
-			fi->next = info;
-		fi = info;
+		n = mrail_count_info(rail_info[i]);
+		if (i == 0)
+			num_info = n;
+		else if (n < num_info)
+			num_info = n;
 
-		/* We only want the first fi_info entry per rail */
-		if (info->next) {
-			fi_freeinfo(info->next);
-			info->next = NULL;
+		num_rails++;
+	}
+
+	if (!num_info) {
+		ret = -FI_ENODATA;
+		goto out;
+	}
+
+	info_array = calloc(num_info, sizeof(struct fi_info *));
+	if (!info_array) {
+		ret = -FI_ENOMEM;
+		goto err;
+	}
+
+	tails = calloc(num_info, sizeof(struct fi_info *));
+	if (!tails) {
+		ret = -FI_ENOMEM;
+		goto err1;
+	}
+
+	/*
+	 * Transpose the info list array:
+	 * 		<rail_info>				<info_array>
+	 * |[rail_1_info_1, rail_1_info_2, ... ]|	|[rail_1_info_1, rail_2_info_1, ... ]|
+	 * |[rail_2_info_1, rail_2_info_2, ... ]| ====>	|[rail_1_info_2, rail_2_info_2, ... ]|
+	 * |		......		        |	|		......		     |
+	 * |[rail_n_info_1, rail_n_info_2, ... ]|	|[rail_1_info_m, rail_2_info_m, ... ]|
+	 */
```

Comment:
I didn't understand the commit message.  Why do this transpose?  It seems counter-intuitive to link infos from unrelated rails together.

Reply:
Each of the linked list of infos from different rails forms a multi-rail info. In the past, only the headers of each rail were linked together, now we want to do the same thing for the 2nd ones, the 3rd ones, ...

Reply:
Can you add the reasoning to the commit message?

Reply:
No problem.